internal/agent.go

```diff
@@ -0,0 +1,96 @@
+package internal
+
+import (
+	"context"
+	"errors"
+	"net"
+	"sync"
+
+	"github.com/dropmorepackets/haproxy-go/pkg/encoding"
+	"github.com/dropmorepackets/haproxy-go/spop"
+	"github.com/rs/zerolog"
+)
+
+type Agent struct {
+	Context      context.Context
+	Applications map[string]*Application
+	Logger       zerolog.Logger
+
+	mtx sync.RWMutex
+}
+
+func (a *Agent) Serve(l net.Listener) error {
+	agent := spop.Agent{
+		Handler:     a,
+		BaseContext: a.Context,
+	}
+
+	return agent.Serve(l)
+}
+
+func (a *Agent) ReplaceApplications(newApps map[string]*Application) {
+	a.mtx.Lock()
+	a.Applications = newApps
+	a.mtx.Unlock()
+}
+
+func (a *Agent) HandleSPOE(ctx context.Context, writer *encoding.ActionWriter, message *encoding.Message) {
+	const (
+		messageCorazaRequest  = "coraza-req"
+		messageCorazaResponse = "coraza-res"
+	)
+
+	var messageHandler func(*Application, context.Context, *encoding.ActionWriter, *encoding.Message) error
+	switch name := string(message.NameBytes()); name {
+	case messageCorazaRequest:
+		messageHandler = (*Application).HandleRequest
+	case messageCorazaResponse:
+		messageHandler = (*Application).HandleResponse
+	default:
+		a.Logger.Debug().Str("message", name).Msg("unknown spoe message")
+		return
+	}
+
+	k := encoding.AcquireKVEntry()
+	defer encoding.ReleaseKVEntry(k)
+	if !message.KV.Next(k) {
+		a.Logger.Panic().Msg("failed reading kv entry")
+		return
+	}
+
+	appName := string(k.ValueBytes())
+	if !k.NameEquals("app") {
+		// Without knowing the app, we cannot continue. We could fall back to a default application,
+		// but all following code would have to support that as we now already read one of the kv entries.
+		a.Logger.Panic().Str("expected", "app").Str("got", appName).Msg("unexpected kv entry")
```

Comment:
I often see the error `unexpected kv entry expected=app got=sample_app` when configuring the app-name `sample_app`. Could this be an issue? I'm not sure why this happens

Reply:
Ah yeah this should not print the field value but the field name

Reply:
Can you check now what the correct error is? The fields in the haproxy/coraza.cfg are expected to be in that specific order

Reply:
Alright. With the new version I see: `unexpected kv entry expected=app got=path`
Have the exact `coraza.cfg` from your examples: (and working most of the time)
```
spoe-message coraza-req
    # Arguments are required to be in this order
    args app=str(sample_app) src-ip=src src-port=src_port dst-ip=dst dst-port=dst_port method=method path=path query=query version=req.ver headers=req.hdrs body=req.body
    event on-frontend-http-request
```

Reply:
Which HAProxy version? I would like to reproduce this.

Reply:
Will E-Mail you the complete config.

Reply:
Found the issue! There was a race inside the decoder. Fixed now :)

Reply:
Nice :+1: 
Changes are up in my env. Will report back tomorrow

Reply:
Have not yet seen that error again.
The only thing that is visible are `reading frame length: read tcp 127.0.0.1:9000->127.0.0.1:xxxxx: read: connection reset by peer` logs.
~1.000 logs in the last 24h

Not sure this is normal. If they are expected - we might want to disable these logs or configure them to be debug-logs.

Reply:
These are expected and happen when HAProxy closes a connection instead of reusing them. These will stay for now as I have to investigate the behaviour of haproxy regarding this, but they aren't anything bad.